Add 1984 -> 1994 completions data (gender only)

cduvallet opened the pull request:
Closes #15 
Closes #42

chrisjkuch commented on scipeds/data/enums.py:
```diff
@@ -48,6 +49,7 @@ class RaceEthn(str, Enum):
     two_or_more = "Two or more races"
     white = "White"
     unknown = "Unknown"
+    total = "Total"
```
It might be better to leave race as null rather than create a new category

cduvallet replied:
Right, I suppose there isn't a "race ethnicity" column for the df when you group by gender, so there isn't a risk of user confusion.

I'm open to either way! Honestly, whichever way reduces the risk of accidentally double-counting -- which leaving it as Null is probably better

cduvallet replied:
Ah, duckdb doesn't allow for NULL values in ENUM columns. So we can either change the race/ethnicity column to a string dtype, or add this "Total" group?

chrisjkuch replied:
There's a slight distinction here - duckdb doesn't allow null values in enum _types_, but null values in a _column with enum values_ are okay. So if we change the query that we use to create the enum _type_ to something like:

```python
create_enum_stmt = f"""
            CREATE TYPE {col}_enum AS ENUM
                (SELECT DISTINCT {col} 
                FROM {COMPLETIONS_TABLE} 
                WHERE {col} IS NOT NULL
                ORDER BY {col})
            """
```

Then we should be able to create the enum type in the way we expect, but still be able to add NULL values to it.

chrisjkuch replied:
Alternatively, we could (accurately) classify them under the existing "unknown" classification.

cduvallet commented:
@chrisjkuch - ready for another re-review. I've gotten the unknown CIPs down to an acceptable level. Here is just for Bachelor's data:

<img width="482" alt="image" src="https://github.com/user-attachments/assets/6f8e2cdb-1598-4f8b-bc1a-f6bb8cb12ed3" />

There are a bunch of CIP 95.000 for 1988 and 1989, but I think these are at the high school or certificate level so I am ok to leave them for now.

chrisjkuch commented on pipeline/cip_crosswalk.py:
```diff
@@ -34,10 +35,65 @@ def __init__(
 
+
+    def _parse_1990_cip_pdf(self):
+        pdf_path = self.crosswalk_dir / "1985-1999" / "91396_cip_crosswalk.pdf"
```
We need to add this to the crosswalk assets, the raw data bucket, and/or a download link

cduvallet replied:
yep, I moved this into the bucket so it'll get downloaded with the other crosswalk assets: `scipeds-data/raw/ipeds_crosswalks/1985-1999`

cduvallet replied:
I ended putting this pdf into the assets bucket because it's actually a _subset_ of the entire raw pdf that I manually created. The raw pdf is still in the scipeds-data bucket (and gets downloaded when you download either from the bucket or from IPEDS); I think it's probably good to leave it in there even if it might be a little bit confusing to have it in two places - but happy to discuss if you think otherwise!

chrisjkuch commented on pipeline/cip_crosswalk.py:
```diff
@@ -113,12 +169,51 @@ def _load_1990_to_2000_crosswalk(self):
 
         self.crosswalk[year_range] = mappers
 
+    def _load_1985_to_1990_crosswalk(self):
```
Something not quite right is happening with the mapping to cip2020 - I end up with string title values in the cip2020 column when I expect 7-digit CIPs.

One way to try to figure out what's going wrong is to see where any values in the mapping have more than 7 characters.

cduvallet replied:
hm, i'm not seeing this issue. it was a problem at first but i fixed it, and now if i run the debugger i'm getting that everything in the cip mapping has 7 characters.

cduvallet replied:
I added a test for this

chrisjkuch reviewed (changes requested):
I reviewed a few things in single comments, summarizing here:

* We should either treat unknown values for race/ethnicity from early years as the existing `unknown` value or as null values, creating the enum appropriately. I think going the `unknown` route will be a bit easier from a processing perspective, and is also accurate.
* Something strange is happening for older (<1990) cip values in certain cases - we should double check that the mapping always gives us a 6-digit CIP.
* We should move pdfs into the `data` folder / make them directly downloadable from the web as well as from cloud storage
* We should rebase this on `main` to incorporate recent changes.

cduvallet commented:
Ready for another review, I think it is done and ready to merge!

> We should either treat unknown values for race/ethnicity from early years as the existing unknown value or as null values, creating the enum appropriately. I think going the unknown route will be a bit easier from a processing perspective, and is also accurate.

Yep, agreed.

> Something strange is happening for older (<1990) cip values in certain cases - we should double check that the mapping always gives us a 6-digit CIP.

I added a test for this; lmk if you are still seeing this issue.

> We should move pdfs into the data folder / make them directly downloadable from the web as well as from cloud storage

- I moved the raw IPEDS data into our bucket, so it gets downloaded with the `make download` command
- I also added & tested code so that all the raw data can be downloaded directly from IPEDS
- The CIP 1990 pdf is in the bucket and a subset of it (appendix B, which is what's used for actually making the crosswalk) is also in the assets folder.

> We should rebase this on main to incorporate recent changes.

Done, though there are still some merge conflicts with main, gahhh whyyyy

chrisjkuch commented on pipeline/cip_crosswalk.py:
```diff
@@ -34,10 +35,70 @@ def __init__(
         self.year_ranges = sorted(list(self.crosswalk.keys()), key=lambda x: x[0])
         self.min_year = min(yr[0] for yr in self.year_ranges)
 
+    def _parse_1990_cip_pdf(self):
```
It seems as though a lot of this is unnecessary now that you've reduced the PDF manually. Do we still need the code that removes chapter and column headers?

cduvallet replied:
Yeah, take a look at the reduced PDF -- it still has "chapters" in it (and a few lines here and there which don't get parsed correctly)

chrisjkuch commented on cip_exploration.ipynb:
A few things here - even though I know we're deleting these notebooks, I want to make sure that once we've made a processed duckdb that all the code here is properly updated / migrated.

* Can you change the local file path to the pdf to refer to the pipeline assets dir?
* Can you change the line that looks for the 1984-1989 cw to the 1984-1991 cw?
* There's a cell that is marked as a code cell that should be a markdown cell

chrisjkuch commented on scipeds/data/enums.py:
```diff
@@ -48,6 +49,7 @@ class RaceEthn(str, Enum):
     two_or_more = "Two or more races"
     white = "White"
     unknown = "Unknown"
+    total = "Total"
```
We should remove this, correct?
```suggestion
```

chrisjkuch commented on scipeds/data/enums.py:
```diff
@@ -37,6 +37,7 @@ class RaceEthn(str, Enum):
         two_or_more (str): Two or more races
         white (str): White
         unknown (str): Unknown
+        total (str): Total
```
```suggestion
```

cduvallet replied:
Good catch!

chrisjkuch commented on scipeds/data/enums.py:
```diff
@@ -458,4 +462,7 @@ class NSFBroadField(str, Enum):
     NCSESFieldGroup.phys_sci: NSFBroadField.phys_earth_sci,
     NCSESFieldGroup.psych: NSFBroadField.soc_behav_sci,
     NCSESFieldGroup.geosci: NSFBroadField.phys_earth_sci,
+    NCSESFieldGroup.unknown: NSFBroadField.unknown,
+    NCSESSciGroup.non_sci: NSFBroadField.non_stem,
+    NCSESSciGroup.unknown: NSFBroadField.unknown,
```
I added this for #42

chrisjkuch reviewed (changes requested):
A few changes to make, plus some changes to address #42 while we're doing pipeline dev.

> Resolve conflicts with main (they seem mostly related to package versioning, can you take a look and see if we should just accept all from main vs this branch? Probably main, right?)

I think I fixed this with a rebase

> Delete my exploration notebooks

Yes, moving these to a separate repo makes sense.

cduvallet commented:
@chrisjkuch - notebooks have been updated and moved to https://github.com/scienceforamerica/sfa-stem/pull/67 , I think all other changes have been addressed!
